File: concepts/Refactoring/Java/Java Refactoring Best Practices/Bloaters/2contrived-complexity/BaseProject/src/main/java/com/onlineshop/CheckoutHandler.java
package com.onlineshop;


import com.onlineshop.items.Item;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDate;
import java.util.List;

public class CheckoutHandler {

    private LocalDate deliveryWindowStart;
    private LocalDate deliveryWindowEnd;

    private double sumItemPrices(List<Item> items){
        double sum = 0;
        for(Item item: items){
            sum += sum + item.price();
        }
        return sum;
    }

    private double applyVoucher(String voucher, double price){
        // check if voucher is valid
        if(isValidVoucher(voucher)){
            price = BigDecimal.valueOf(price * 0.95).setScale(2, RoundingMode.HALF_EVEN).doubleValue();
        } else {
            System.out.println("Voucher invalid");
        }
        return price;
    }

    private boolean isValidVoucher(String voucher){
        return voucher.equals("GIMME_DISCOUNT") || voucher.equals("CHEAPER_PLEASE");
    }

    private double addDeliveryFee(String membership, String address, double price){
        // handle delivery fee
        if(isEligibleForFreeDelivery(membership)){
            // do nothing
        } else {
            if(isUSAddress(address)){
                System.out.println("Adding flat delivery fee of 5 USD");
                price = price + 5;
            } else {
                System.out.println("Adding flat global delivery fee of 10 USD");
                price = price + 10;
            }
        }
        return price;
    }
    private boolean isEligibleForFreeDelivery(String membership){
        return  membership.equalsIgnoreCase("GOLD");
    }

    private boolean isUSAddress(String address){
        return address.contains("US");
    }

    public double calculateTotal(List<Item> items, String voucher, String membership, String address){
        double baseTotal = sumItemPrices(items);
        baseTotal = applyVoucher(voucher, baseTotal);
        baseTotal = addDeliveryFee(membership, address, baseTotal);
        return baseTotal;
    }



    public void setDeliveryTimeWindow(LocalDate deliveryStart, LocalDate deliveryEnd){
        this.deliveryWindowStart = deliveryStart;
        this.deliveryWindowEnd = deliveryEnd;

        System.out.println(String.format("Your Order will delivered some time between %s and %s", deliveryWindowStart, deliveryWindowEnd));
    }

}

File: software-engineering-concepts/Refactoring/PHP/PHP Refactoring Best Practices/1Bloaters/0BaseProject/src/CheckoutHandler.php
<?php 
namespace app\handlers;

use Exception;

class CheckoutHandler 
{

    private $deliveryWindowStart;
    private $deliveryWindowEnd;

    public function calculateTotal($items, $voucher, $membership, $address)
    {

        $baseTotal = 0;

        // sum up the prices
        $prices = array();
        foreach($items as $item){
            $prices[] = $item->price();
        }

        foreach($prices  as $price){
            $baseTotal = $baseTotal + $price;
        }

        // check if voucher is valid
        if($voucher === "GIMME_DISCOUNT" || $voucher === "CHEAPER_PLEASE"){
             $baseTotal = $baseTotal * 0.95;
        } else {
            // throw new Exception("Voucher invalid");
            echo "Voucher invalid";

        }

        // handle delivery fee
        if(strtoupper($membership) === "GOLD"){
            // do nothing
        } else {
            if(preg_match("/US$/", $address)){
                echo "Adding flat delivery fee of 5 USD";
                $baseTotal = $baseTotal + 5;
            } else {
                echo "Adding flat global delivery fee of 10 USD";
                $baseTotal = $baseTotal + 10;
            }
        }
        return $baseTotal;
    }



    public function setDeliveryTimeWindow($deliveryStart, $deliveryEnd){
        $this->deliveryWindowStart = $deliveryStart;
        $this->deliveryWindowEnd = $deliveryEnd;
        echo"Your Order will delivered some time between $this->deliveryWindowStart and $this->deliveryWindowEnd \n";
    }

}
